lib/mongo-milestone-db.js

```diff
@@ -79,19 +87,52 @@ class MongoMilestoneDB extends MilestoneDB {
       });
   }
 
-  _getMilestoneSnapshot(collectionName, id, version) {
-    return this._collection(collectionName)
-      .then((collection) => {
-        const query = {d: id};
-        if (version != null) {
-          query.v = {$lte: version};
-        }
+  _getMilestoneSnapshotByVersion(collectionName, id, version) {
+    if (!id) return Promise.reject(new InvalidIdError());
+    if (!this._isValidVersion(version)) {
+      return Promise.reject(new InvalidVersionError());
+    }
+
+    const query = {d: id};
+    if (version != null) {
+      query.v = {$lte: version};
+    }
 
-        const sort = {v: -1};
+    const options = {
+      sort: {v: -1},
+    };
 
-        return collection
-          .findOne(query, {sort: sort});
-      })
+    return this._getMilestoneSnapshotByQuery(collectionName, query, options);
+  }
+
+  _getMilestoneSnapshotByTimestamp(collectionName, id, timestamp, isAfterTimestamp) {
+    if (!id) return Promise.reject(new InvalidIdError());
+    if (!this._isValidTimestamp(timestamp)) {
+      return Promise.reject(new InvalidTimestampError());
+    }
+
+    const nullSortOrder = isAfterTimestamp ? -1 : 1;
+
+    const query = {d: id};
+    const options = {
+      limit: 1,
+      sort: {'m.mtime': nullSortOrder},
+    };
+
+    if (timestamp !== null) {
```

Comment:
I think it would be easier to read this code if we didn't manipulate values once set. This is a bit tricky to follow because of the overrides and negation.

How about something like:

```js
const limit = 1;
const query, sortOrder;
if (timestamp === null) {
  query = {d: id};
  sortOrder = (isAfterTimestamp) ? -1 : 1;
} else {
  const comparator = (isAfterTimestamp) ? {$gte: timestamp} : {$lte: timestamp};
  query = {
    d: id,
    'm.mtime': comparator
  };
  sortOrder = (isAfterTimestamp) ? 1 : -1;
}
const options = {
  limit: limit,
  sort: {'m.mtime': sortOrder},
};
```

Reply:
Looks good to me. Pushing.